sdk/src/client/secret/mod.rs

```diff
@@ -373,99 +410,103 @@ impl SecretManager {
-                    Address::Alias(alias_output.alias_address(input.output_id())),
-                    current_block_index,
-                ),
-                Output::Nft(nft_output) => block_indexes.insert(
-                    Address::Nft(nft_output.nft_address(input.output_id())),
-                    current_block_index,
-                ),
-                _ => None,
-            };
+                let chain = input.chain.as_ref().ok_or(Error::MissingBip32Chain)?;
+
+                let block = secret_manager.signature_unlock(&hashed_essence, chain).await?;
+                blocks.push(block);
+
+                // Add the ed25519 address to the block_indexes, so it gets referenced if further inputs have
+                // the same address in their unlock condition
+                block_indexes.insert(input_address, current_block_index);
+            }
         }
 
-        Ok(Unlocks::new(blocks)?)
+        // When we have an alias or Nft output, we will add their alias or nft address to block_indexes,
+        // because they can be used to unlock outputs via [Unlock::Alias] or [Unlock::Nft],
+        // that have the corresponding alias or nft address in their unlock condition
+        match &input.output {
+            Output::Alias(alias_output) => block_indexes.insert(
+                Address::Alias(alias_output.alias_address(input.output_id())),
+                current_block_index,
+            ),
+            Output::Nft(nft_output) => block_indexes.insert(
+                Address::Nft(nft_output.nft_address(input.output_id())),
+                current_block_index,
+            ),
+            _ => None,
+        };
     }
 
-    /// Sign a transaction
-    pub async fn sign_transaction(
-        &self,
-        prepared_transaction_data: PreparedTransactionData,
-    ) -> crate::client::Result<Payload> {
-        log::debug!("[sign_transaction] {:?}", prepared_transaction_data);
-        let current_time = unix_timestamp_now().as_secs() as u32;
+    Ok(Unlocks::new(blocks)?)
+}
 
-        let unlocks = self
-            .sign_transaction_essence(&prepared_transaction_data, Some(current_time))
-            .await?;
-        let tx_payload = TransactionPayload::new(prepared_transaction_data.essence.clone(), unlocks)?;
+pub(crate) async fn default_sign_transaction<M: SecretManage>(
+    secret_manager: &M,
+    prepared_transaction_data: PreparedTransactionData,
+) -> crate::client::Result<Payload>
+where
+    crate::client::Error: From<M::Error>,
+{
+    log::debug!("[sign_transaction] {:?}", prepared_transaction_data);
+    let current_time = unix_timestamp_now().as_secs() as u32;
 
-        validate_transaction_payload_length(&tx_payload)?;
+    let unlocks = secret_manager
+        .sign_transaction_essence(&prepared_transaction_data, Some(current_time))
+        .await?;
+    let tx_payload = TransactionPayload::new(prepared_transaction_data.essence.clone(), unlocks)?;
```

Comment:
Can't we destructure `PreparedTransactionData`, so we don't need to clone the essence here?